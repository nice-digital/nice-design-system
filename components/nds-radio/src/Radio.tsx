import React, { Component } from "react";
import classnames from "classnames";
import "../scss/radio.scss";

export interface RadioProps {
	[prop: string]: unknown;
	disabled?: boolean;
	error?: boolean | string;
	name?: string;
	inline?: boolean;
	label?: React.ReactNode;
	value: string;
	hint?: string;
}

export const Radio: React.FC<RadioProps> = (props: RadioProps) => {
	const { disabled, hint, error, name, label, value, inline, ...rest } = props;
	if (!value || value === "") return null;
	const classNames = classnames({
		radio: true,
		"radio--inline": inline,
		"radio--error": error
	});
	const unique = `${name}_${value}`;

	return (
		<>
			{error && error.toString().length && (
				<p className="radio__error-message">{error}</p>
			)}
			<div className={classNames}>
				<input
					disabled={disabled}
					className="radio__input"
					name={name}
					type="radio"
					id={unique}
					{...rest}
				/>
				<label className="radio__label" htmlFor={unique}>
					{label ? label : value}
				</label>
				{hint && <span className="checkbox__hint">{hint}</span>}
			</div>
		</>
	);
};
