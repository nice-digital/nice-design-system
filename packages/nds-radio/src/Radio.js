import React, { Component } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import "../scss/radio.scss";

export class Radio extends Component {
	render() {
		const {
			disabled,
			hint,
			error,
			name,
			label,
			value,
			inline,
			...rest
		} = this.props;
		if (!value || value === "") return null;
		const classNames = classnames({
			radio: true,
			"radio--inline": inline,
			"radio--error": error
		});
		const unique = `${name}_${value}`;
		return (
			<>
				{error && error.length && (
					<p className="radio__error-message">{error}</p>
				)}
				<div className={classNames}>
					<input
						disabled={disabled}
						className="radio__input"
						name={name}
						type="radio"
						id={unique}
						{...rest}
					/>
					<label className="radio__label" htmlFor={unique}>
						{label ? label : value}
					</label>
					{hint && <span className="checkbox__hint">{hint}</span>}
				</div>
			</>
		);
	}
}

Radio.propTypes = {
	disabled: PropTypes.bool,
	error: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
	name: PropTypes.string,
	inline: PropTypes.bool,
	label: PropTypes.node,
	value: PropTypes.string.isRequired,
	hint: PropTypes.string
};
