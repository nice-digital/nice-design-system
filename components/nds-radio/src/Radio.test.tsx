import React from "react";
import { render } from "@testing-library/react";

import { Radio, RadioProps } from "./Radio";

const props: RadioProps = {
	label: "Yes, please.",
	value: "yes",
	disabled: false,
	error: false,
	inline: false,
	name: "radio-group"
};

describe("Radio", () => {
	it("should match snapshot with some default attributes", () => {
		const wrapper = render(<Radio {...props} />);
		expect(wrapper).toMatchSnapshot();
	});

	it("should match snapshot with hint text supplied", () => {
		const wrapper = render(
			<Radio {...props} hint="This is a hint to appear." />
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("should not render a radio control if it doesn't have a value", () => {
		const localProps = Object.assign({}, props);
		localProps.value = "";
		const wrapper = render(<Radio {...localProps} />);
		expect(wrapper.queryByRole("radio")).toBe(null);
	});

	it("should prevent actions if the button is disabled", () => {
		const localProps = Object.assign({}, props);
		localProps.disabled = true;
		const wrapper = render(<Radio {...localProps} />);
		const input: HTMLInputElement = wrapper.getByRole(
			"radio"
		) as HTMLInputElement;
		expect(input.disabled).toBe(true);
	});

	it("should add an error class to the parent if the error prop is true", () => {
		const localProps = Object.assign({}, props);
		localProps.error = true;
		const { container } = render(<Radio {...localProps} />);
		const parent = container.querySelector("div.radio--error");
		expect(parent).toBeInTheDocument();
	});

	it("should display error text above the control if an error string is supplied", () => {
		const localProps = Object.assign({}, props);
		localProps.error = "Error message here.";
		const { container } = render(<Radio {...localProps} />);
		const parent = container.querySelector("p.radio__error-message");
		expect(parent?.textContent).toBe("Error message here.");
	});

	it("should pass through any additional attributes supplied", () => {
		const localProps = Object.assign({}, props);
		localProps.defaultChecked = true;
		localProps["data-something"] = "test";
		const wrapper = render(<Radio {...localProps} />);
		const checked: HTMLInputElement = wrapper.getByRole(
			"radio"
		) as HTMLInputElement;
		expect(checked.checked).toBe(true);
		expect(checked.getAttribute("data-something")).toBe("test");
	});
});
