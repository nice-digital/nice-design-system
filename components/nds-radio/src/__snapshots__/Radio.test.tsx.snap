// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio should match snapshot with hint text supplied 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="radio"
      >
        <input
          class="radio__input"
          id="radio-group_yes"
          name="radio-group"
          type="radio"
        />
        <label
          class="radio__label"
          for="radio-group_yes"
        >
          Yes, please.
        </label>
        <span
          class="checkbox__hint"
        >
          This is a hint to appear.
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="radio"
    >
      <input
        class="radio__input"
        id="radio-group_yes"
        name="radio-group"
        type="radio"
      />
      <label
        class="radio__label"
        for="radio-group_yes"
      >
        Yes, please.
      </label>
      <span
        class="checkbox__hint"
      >
        This is a hint to appear.
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Radio should match snapshot with some default attributes 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="radio"
      >
        <input
          class="radio__input"
          id="radio-group_yes"
          name="radio-group"
          type="radio"
        />
        <label
          class="radio__label"
          for="radio-group_yes"
        >
          Yes, please.
        </label>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="radio"
    >
      <input
        class="radio__input"
        id="radio-group_yes"
        name="radio-group"
        type="radio"
      />
      <label
        class="radio__label"
        for="radio-group_yes"
      >
        Yes, please.
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
